### PR TITLE
mops benchmark

### PR DESCRIPTION
**Pull request by AndyGura**
Benchmark results:

```
Sha256

Hash various message lengths from different types of input. Blocks are 64 bytes.


Instructions

|           |      0 |  1k blocks |        1 MB |
| :-------- | -----: | ---------: | ----------: |
| fromBlob  | 15_175 | 17_698_623 | 289_734_090 |
| fromArray | 15_545 | 20_258_993 | 331_677_500 |
| fromIter  | 15_757 | 18_083_205 | 296_026_128 |


Heap

|           |   0 | 1k blocks | 1 MB |
| :-------- | --: | --------: | ---: |
| fromBlob  | 228 |       228 |  228 |
| fromArray | 228 |       228 |  228 |
| fromIter  | 228 |       228 |  228 |


——————————————————————————————————————————————————

Running bench/sha512.bench.mo...



Sha512

Hash various message lengths from different types of input. Blocks are 128 bytes.


Instructions

|           |      0 |  1k blocks |        1 MB |
| :-------- | -----: | ---------: | ----------: |
| fromBlob  | 21_024 | 30_412_922 | 248_988_584 |
| fromArray | 21_394 | 35_533_292 | 290_931_749 |
| fromIter  | 21_606 | 31_181_479 | 255_280_377 |


Heap

|           |   0 | 1k blocks | 1 MB |
| :-------- | --: | --------: | ---: |
| fromBlob  | 228 |       228 |  228 |
| fromArray | 228 |       228 |  228 |
| fromIter  | 228 |       228 |  228 |
```

### CODE REVIEW

File: bench/sha256.bench.mo [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      rows.size() * cols.size(),
+      func(i) {
+        let row : Nat = i % rows.size();
+        let col : Nat = i / rows.size();
+
+        let source = rowSourceArrays[col];
+
+        switch (row) {
+          case (0) {
+            let blob = Blob.fromArray(source);
```
Comment (timohanke) [inline bench/sha256.bench.mo:48]: This conversion should happen in advance, outside of the measured code.

But interesting to see that fromBlob is faster than fromArray even when this conversion is included. I did not expect that.

File: bench/sha256.bench.mo [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            func() = ignore Sha256.fromBlob(#sha256, blob);
+          };
+          case (1) {
+            func() = ignore Sha256.fromArray(#sha256, source);
+          };
+          case (2) {
+            let iter = Iter.fromArray(source);
```
Comment (timohanke) [inline bench/sha256.bench.mo]: This is insignificant amount of work. But just for consistency with the blob test above (for the reader) I would create this iter also outside the measured code.

**Review by timohanke — COMMENTED** [review]
Left 1 change request

File: bench/sha256.bench.mo (hunk 1 above)
Comment (AndyGura) [reply bench/sha256.bench.mo:48]: It happens outside of measured code. See that next line returns function, which will be measured

File: bench/sha256.bench.mo (hunk 2 above)
Comment (AndyGura) [reply bench/sha256.bench.mo]: It's also should not be included in the measurements

**Comment by timohanke** [comment]
A second change request. Currently the fromIter and fromArray lead to the exact same code (that's how internally fromArray works). So a more interesting test would be to change the fromIter test and provide a "faster" iter which is not based on an Array but you construct it directly where the `next()` is a constant function returning `0x5f` (plus a counter so that it terminates).

**Comment by AndyGura** [comment]
@timohanke changed iterator in `toIter` benchmark. Stats in PR description updated

File: bench/sha256.bench.mo (hunk 1 above)
Comment (timohanke) [reply bench/sha256.bench.mo:48]: I see. Got it

File: bench/sha256.bench.mo (hunk 2 above)
Comment (timohanke) [reply bench/sha256.bench.mo]: Got it. Thanks

**Review by timohanke — APPROVED** [review]
LGTM

**Pull request merged**